File: Project_Sourcecode_java/src/main/java/edu/iit/cs586/project/state/State.java
package edu.iit.cs586.project.state;

import edu.iit.cs586.project.MDA_EFSM;

/**
 * @author CharmingLee <charminglee@jieztech.com>
 * @date 2022/5/1 22:02
 */
public abstract class State {
    protected MDA_EFSM m;

    public State(MDA_EFSM m) {
        this.m = m;
    }

    public String getName() {
        return "State";
    }
    
    private String getMsg() {
        return "WARN: The current state does not support this operation";
    }

    public void login() {
        System.out.println(getMsg());
    }

    public void incorrectLogin() {
        System.out.println(getMsg());
    }

    public void open() {
        System.out.println(getMsg());
    }

    public void incorrectPin(int max) {
        System.out.println(getMsg());
    }

    public void correctPinAboveMin() {
        System.out.println(getMsg());
    }

    public void logout() {
        System.out.println(getMsg());
    }

    public void noFunds() {
        System.out.println(getMsg());
    }

    public void suspend() {
        System.out.println(getMsg());
    }

    public void deposit() {
        System.out.println(getMsg());
    }

    public void incorrectLock() {
        System.out.println(getMsg());
    }

    public void lock() {
        System.out.println(getMsg());
    }

    public void withdraw() {
        System.out.println(getMsg());
    }

    public void balance() {
        System.out.println(getMsg());
    }

    public void aboveMinBalance() {
        System.out.println(getMsg());
    }

    public void belowMinBalance() {
        System.out.println(getMsg());
    }

    public void withdrawBelowMinBalance() {
        System.out.println(getMsg());
    }

    public void unlock() {
        System.out.println(getMsg());
    }

    public void incorrectUnlock() {
        System.out.println(getMsg());
    }

    public void correctPinBelowMin() {
        System.out.println(getMsg());
    }

    public void close() {
        System.out.println(getMsg());
    }

    public void activate() {
        System.out.println(getMsg());
    }
}

File: Project_Sourcecode_go/state/state.go
package state

import (
	"fmt"
)

type State struct {
	M *MdaEfsm
}

type StateApi interface {
	Login()
	IncorrectLogin()
	Open()
	IncorrectPin(max int)
	CorrectPinAboveMin()
	Logout()
	NoFunds()
	Suspend()
	Deposit()
	IncorrectLock()
	Lock()
	Withdraw()
	Balance()
	AboveMinBalance()
	BelowMinBalance()
	WithdrawBelowMinBalance()
	Unlock()
	IncorrectUnlock()
	CorrectPinBelowMin()
	Close()
	Activate()
	GetName() string
}

func (s *State) Open() {
	fmt.Println("current state not support this function")
}
func (s *State) Login() {
	fmt.Println("current state not support this function")
}
func (s *State) IncorrectLogin() {
	fmt.Println("current state not support this function")
}
func (s *State) IncorrectPin(max int) {
	fmt.Println("current state not support this function")
}
func (s *State) CorrectPinAboveMin() {
	fmt.Println("current state not support this function")
}
func (s *State) Logout() {
	fmt.Println("current state not support this function")
}
func (s *State) NoFunds() {
	fmt.Println("current state not support this function")
}
func (s *State) Suspend() {
	fmt.Println("current state not support this function")
}
func (s *State) Deposit() {
	fmt.Println("current state not support this function")
}
func (s *State) IncorrectLock() {
	fmt.Println("current state not support this function")
}
func (s *State) Lock() {
	fmt.Println("current state not support this function")
}
func (s *State) Withdraw() {
	fmt.Println("current state not support this function")
}
func (s *State) Balance() {
	fmt.Println("current state not support this function")
}
func (s *State) AboveMinBalance() {
	fmt.Println("current state not support this function")
}
func (s *State) BelowMinBalance() {
	fmt.Println("current state not support this function")
}
func (s *State) WithdrawBelowMinBalance() {
	fmt.Println("current state not support this function")
}
func (s *State) Unlock() {
	fmt.Println("current state not support this function")
}
func (s *State) IncorrectUnlock() {
	fmt.Println("current state not support this function")
}
func (s *State) CorrectPinBelowMin() {
	fmt.Println("current state not support this function")
}
func (s *State) Close() {
	fmt.Println("current state not support this function")
}
func (s *State) Activate() {
	fmt.Println("current state not support this function")
}

func (s *State) GetName() string {
	return "state"
}
